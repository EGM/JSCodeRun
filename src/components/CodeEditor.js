import React from 'react';
import { TextInput, Text, StyleSheet } from 'react-native';
import * as reservedWords from '../constants/reservedWords';

const styles = StyleSheet.create({
  basicKeywordStyle: {
    color: '#769aa5'
  },
  funcKeywordStyle: {
    color: '#ecf13a'
  },
  advancedKeywordStyle: {
    color: '#a1edd5'
  },
  stringStyle: {
    color: '#cc7832'
  },
  numberStyle: {
    color: '#a5c25c'
  },
  returnStyle: {
    color: '#9876aa'
  },
  defaultStyle: {
    color: '#eee'
  }
});

const Wrapper = ({ text }) => {
  const words = text.split(
    new RegExp(`([0-9]+|${reservedWords.allWords.join('|')}|".*?"|'.*?'|\`.*?\`)`, 'g')
  );

  const styleWrappedTexts = words.map((word, i) => {
    if (!word) {
      return null;
    }

    if (reservedWords.basicWords.indexOf(word) >= 0) {
      partStyle = styles.basicKeywordStyle;
    } else if (reservedWords.funcWords.indexOf(word) >= 0) {
      partStyle = styles.funcKeywordStyle;
    } else if (reservedWords.advancedWords.indexOf(word) >= 0) {
      partStyle = styles.advancedKeywordStyle;
    } else if (!isNaN(parseInt(word))) {
      partStyle = styles.numberStyle;
    } else if (
      (word.startsWith('"') && word.endsWith('"')) ||
      (word.startsWith("'") && word.endsWith("'")) ||
      (word.startsWith('`') && word.endsWith('`'))
    ) {
      partStyle = styles.stringStyle;
    } else {
      partStyle = styles.defaultStyle;
    }
    return (
      <Text key={i} style={partStyle}>
        {word}
      </Text>
    );
  });
  return <Text>{styleWrappedTexts}</Text>;
};

class CodeEditor extends React.Component {
  render() {
    return (
      <TextInput
        style={{
          height: '100%',
          backgroundColor: '#222',
          color: 'white'
        }}
        autoCapitalize="none"
        autoCorrect={false}
        multiline
        onChangeText={this.props.handleCodeChange}
      >
        <Wrapper text={this.props.code} />
      </TextInput>
    );
  }
}
export default CodeEditor;
