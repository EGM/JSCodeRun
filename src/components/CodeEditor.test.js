import React, { Component } from 'react';
import renderer from 'react-test-renderer';
import CodeEditor from './CodeEditor';

describe('CodeEditor', () => {
  it('should render correctly with code', () => {
    const rendered = renderer.create(
      <CodeEditor 
        handleCodeChange={() => {}}
        code="console.log('Hello');"/>
    ).toJSON();
    expect(rendered).toMatchSnapshot();
  });
})