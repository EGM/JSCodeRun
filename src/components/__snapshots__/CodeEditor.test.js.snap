// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodeEditor should render correctly with code 1`] = `
<TextInput
  autoCapitalize="none"
  autoCorrect={false}
  blurOnSubmit={false}
  keyboardType="ascii-capable"
  multiline={true}
  onChangeText={[Function]}
  style={
    Object {
      "backgroundColor": "#222",
      "color": "white",
      "fontSize": 15,
      "height": "100%",
      "padding": 5,
      "textAlignVertical": "top",
    }
  }
  underlineColorAndroid="transparent"
>
  <Text
    accessible={true}
    allowFontScaling={true}
    disabled={false}
    ellipsizeMode="tail"
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#eee",
        }
      }
    >
      
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#71d5c3",
        }
      }
    >
      console
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#eee",
        }
      }
    >
      
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#769aa5",
        }
      }
    >
      .
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#eee",
        }
      }
    >
      log
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#769aa5",
        }
      }
    >
      (
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#eee",
        }
      }
    >
      
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#cc7832",
        }
      }
    >
      'Hello'
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#eee",
        }
      }
    >
      
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#769aa5",
        }
      }
    >
      )
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#eee",
        }
      }
    >
      ;
    </Text>
  </Text>
</TextInput>
`;
